src/ui/marker.js

```diff
@@ -534,7 +536,12 @@ export default class Marker extends Evented {
         // we only round them when _update is called with `moveend` or when its called with
         // no arguments (when the Marker is initialized or Marker#setLngLat is invoked).
         if (!e || e.type === "moveend") {
-            this._pos = this._pos.round();
+            this._updateFrameId = window.requestAnimationFrame(() => {
+                if (this._element && this._pos && this._anchor) {
+                    this._pos = this._pos.round();
+                    DOM.setTransform(this._element, `${anchorTranslate[this._anchor]} translate(${this._pos.x}px, ${this._pos.y}px) ${pitch} ${rotation}`);
+                }
+            });
```

Comment:
It's true that we don't really want postponed snapping on `move` events, but I'm not sure if that would actually cause an issue. All `move` events should be followed by either another `move` or `moveend` event, so I think the delayed rounding should be canceled by your call to  `window.cancelAnimationFrame`. There might be cases (maybe due to latency) where the event isn't called in time, so this needs testing. (If it proves to be an issue, I think we could solve it by adding a third option for `move` events to do no rounding at all)

The other way sounds similar to our discussion on https://github.com/mapbox/mapbox-gl-js/pull/11136. I think it's a viable idea and I like its simplicity in the code, but as @mourner mentioned we prefer to avoid adding extra user options unless there is no alternative. Ideally, we want this to work without developers even having to consider it.

> My main reason for proposing this option is to avoid expanding the API surface for something that can be an internal detail users don't have to think about. If we can make it to _just work_ without any user-side changes for both cases, that's ideal.

My preference would be to continue this current PR, and keep the extra option idea as a backup if we're sure that this doesn't meet your needs. What do you think? Is there something I'm missing about why the other approach might better meet your use case?

With either approach, I think we could simplify the code by changing `_update()` to take in a boolean instead of `e?: {type: 'move' | 'moveend'}`

Reply:
Thank you @SnailBones, Yes it's clear.
Ok for this PR should I change anything?

Reply:
Thank you @malekeym!
Yes, do the changes I mention [above](https://github.com/mapbox/mapbox-gl-js/pull/11167#discussion_r736046490) make sense to you? Let me know if you need more clarification.
This will also need some unit tests before we can merge it in.

Reply:
@SnailBones Yes, it's clear, I think changing `_update` to accept `boolean` is better, I change the code, I'll appreciate to check changes and let me know if something is missed or wrong, if all things are ok, let me know to add unit tests. Thank you.

Reply:
Great work! This looks good to me and I think it's ready for unit tests.

Reply:
Thank you @SnailBones I add unit tests, too.

Reply:
This is a great start @malekeym! Can you also add tests to ensure that the snap is _not_ happening immediately on `setLngLat` on map move?

Reply:
@SnailBones I add some unit tests for `move` and `moveend` and `setLngLat`, could you check, and if another cases that my test not cover let me know.